hubblestack/loader.py

```diff
@@ -0,0 +1,1292 @@
+import importlib.util
+
+import pkg_resources
+
+log = logging.getLogger(__name__)
+
+SALT_BASE_PATH = os.path.abspath(salt.syspaths.INSTALL_DIR)
+HUBBLE_BASE_PATH = os.path.abspath(hubblestack.syspaths.INSTALL_DIR)
+LOADED_BASE_NAME = 'hubble.loaded'
+
+MODULE_KIND_SOURCE = 1
+MODULE_KIND_COMPILED = 2
+MODULE_KIND_EXTENSION = 3
+MODULE_KIND_PKG_DIRECTORY = 5
+SUFFIXES = []
+for suffix in importlib.machinery.EXTENSION_SUFFIXES:
+    SUFFIXES.append((suffix, 'rb', MODULE_KIND_EXTENSION))
+for suffix in importlib.machinery.SOURCE_SUFFIXES:
+    SUFFIXES.append((suffix, 'rb', MODULE_KIND_SOURCE))
+for suffix in importlib.machinery.BYTECODE_SUFFIXES:
+    SUFFIXES.append((suffix, 'rb', MODULE_KIND_COMPILED))
+MODULE_KIND_MAP = {
+    MODULE_KIND_SOURCE: importlib.machinery.SourceFileLoader,
+    MODULE_KIND_COMPILED: importlib.machinery.SourcelessFileLoader,
+    MODULE_KIND_EXTENSION: importlib.machinery.ExtensionFileLoader
+}
+
+PY3_PRE_EXT = \
+    re.compile(r'\.cpython-{0}{1}(\.opt-[1-9])?'.format(*sys.version_info[:2]))
+
+# Will be set to pyximport module at runtime if cython is enabled in config.
+pyximport = None
+
+def _module_dirs(
+        opts,
+        ext_type,
+        tag=None,
+        int_type=None,
+        ext_dirs=True,
+        ext_type_dirs=None,
+        base_path=None,
+        ):
+
+    if tag is None:
+        tag = ext_type
+
+    ext_types = os.path.join(opts['extension_modules'], ext_type)
+    sys_types = os.path.join(base_path or HUBBLE_BASE_PATH, int_type or ext_type)
+
+    # XXX should be removed eventually:
+    old_types = os.path.join(SALT_BASE_PATH, int_type or ext_type)
+
+    ext_type_types = []
+    if ext_dirs:
+        if ext_type_dirs is None:
+            ext_type_dirs = '{0}_dirs'.format(tag)
+        if ext_type_dirs in opts:
+            ext_type_types.extend(opts[ext_type_dirs])
+        for entry_point in pkg_resources.iter_entry_points('salt.loader', ext_type_dirs):
+            try:
+                loaded_entry_point = entry_point.load()
+                for path in loaded_entry_point():
+                    ext_type_types.append(path)
+            except Exception as exc:
+                log.error("Error getting module directories from %s: %s", _format_entrypoint_target(entry_point), exc)
+                log.debug("Full backtrace for module directories error", exc_info=True)
+
+    cli_module_dirs = []
+    # The dirs can be any module dir, or a in-tree _{ext_type} dir
+    for _dir in opts.get('module_dirs', []):
+        # Prepend to the list to match cli argument ordering
+        maybe_dir = os.path.join(_dir, ext_type)
+        if os.path.isdir(maybe_dir):
+            cli_module_dirs.insert(0, maybe_dir)
+            continue
+
+        maybe_dir = os.path.join(_dir, '_{0}'.format(ext_type))
+        if os.path.isdir(maybe_dir):
+            cli_module_dirs.insert(0, maybe_dir)
+
+    return cli_module_dirs + ext_type_types + [ext_types, sys_types, old_types]
+
+
+def modules(
```

Comment:
Its a rename from minion_mod(), this change should also be done in daemon.py, it still uses minion_mods()

Reply:
I can do that. I provided the alias so they'll both work for now. But we should certainly aim to remove the alias eventually.